### PR TITLE
Fix for security issue spotted by George Agapov

### PR DESCRIPTION
**Pull request by jspada**
`fiat_pasta_fp_sqrt`'s Tonelli--Shanks implementation doesn't terminate if `value` is not a square.

We can check for this with `value^{(p-1)/2} != 1`

### CODE REVIEW

File: pasta_fp.c [hunk 1]
```diff
@@ -2152,6 +2152,12 @@ void fiat_pasta_fp_sqrt(uint64_t x[4], const uint64_t value[4]) {
       return;
     }
 
+    uint64_t check[4];
+    fiat_pasta_fp_pow(check, value, P_MINUS_1_OVER_2, P_MINUS_1_OVER_2_LEN);
```
Comment (imeckler) [inline pasta_fp.c:2156]: should we have it write an error value in the result?

Comment (jspada) [reply pasta_fp.c:2156]: Looks like it's used only in one place!  Definitely or just return success or failure.

File: crypto.c [hunk 2]
```diff
@@ -887,7 +887,7 @@ void compress(Compressed *compressed, const Affine *pt) {
   compressed->is_odd = y_bigint[0] & 1;
 }
 
-void decompress(Affine *pt, const Compressed *compressed) {
+bool decompress(Affine *pt, const Compressed *compressed) {
```
Comment (mimoo) [inline crypto.c:890]: what about x coordinates that are not on the curve?

File: crypto.c [hunk 3]
```diff
@@ -898,7 +898,9 @@ void decompress(Affine *pt, const Compressed *compressed) {
   fiat_pasta_fp_add(y2, x3, GROUP_COEFF_B);
 
   Field y_pre;
-  fiat_pasta_fp_sqrt(y_pre, y2);
+  if (!fiat_pasta_fp_sqrt(y_pre, y2)) {
+    return false;
+  }
   Field y_pre_bigint;
   fiat_pasta_fp_from_montgomery(y_pre_bigint, y_pre);
```
Comment (mimoo) [inline crypto.c:905]: how does this work when we never converted to_montgomery in the first place?

File: pasta_fp.c [hunk 4]
```diff
@@ -2144,12 +2144,18 @@ void fiat_pasta_fp_print(const uint64_t x[4]) {
...
+bool fiat_pasta_fp_sqrt(uint64_t x[4], const uint64_t value[4]) {
     // A few assertions to make sure s, t, and nqr are initialized.
 
     if (fiat_pasta_fp_equals_zero(value)) {
       for (size_t j = 0; j < 4; ++j) { x[j] = 0; }
-      return;
+      return true;
+    }
+
+    uint64_t check[4];
```
Comment (mimoo) [inline pasta_fp.c:2155]: shouldn't you be using the types you defined elsewhere as `typedef uint64_t Field[LIMBS_PER_FIELD];`

File: pasta_fp.c [hunk 5]
```diff
@@ -17,13 +17,19 @@
...
+#include <stddef.h>
+
 typedef unsigned char fiat_pasta_fp_uint1;
 typedef signed char fiat_pasta_fp_int1;
 
 #if (-1 & 3) != 3
 #error "This code only works on a two's complement system"
 #endif
 
+// x^{(p - 1) / 2}
```
Comment (mimoo) [inline pasta_fp.c:30]: why is there an `x` in this comment?

File: pasta_fp.c [hunk 6]
```diff
@@ -17,13 +17,19 @@
...
+const bool P_MINUS_1_OVER_2[] = {1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+const size_t P_MINUS_1_OVER_2_LEN = 254;
```
Comment (mimoo) [inline pasta_fp.c:32]: can't you just do `sizeof(P_MINUS_1_OVER_2)`?

File: crypto.c (hunk 2 above)
Comment (mimoo) [reply crypto.c:890]: I guess if x is not on the curve, x^3 + b wouldn't be a square then!

File: pasta_fp.c [hunk 7]
```diff
@@ -2144,12 +2144,18 @@ void fiat_pasta_fp_print(const uint64_t x[4]) {
...
 
-void fiat_pasta_fp_sqrt(uint64_t x[4], const uint64_t value[4]) {
+bool fiat_pasta_fp_sqrt(uint64_t x[4], const uint64_t value[4]) {
     // A few assertions to make sure s, t, and nqr are initialized.
 
     if (fiat_pasta_fp_equals_zero(value)) {
       for (size_t j = 0; j < 4; ++j) { x[j] = 0; }
-      return;
+      return true;
+    }
```
Comment (mimoo) [inline pasta_fp.c:2153]: looks like this function is not used anywhere?

File: pasta_fp.c [hunk 8]
```diff
@@ -2144,12 +2144,18 @@ void fiat_pasta_fp_print(const uint64_t x[4]) {
...
+
+    uint64_t check[4];
+    fiat_pasta_fp_pow(check, value, P_MINUS_1_OVER_2, P_MINUS_1_OVER_2_LEN);
+    if (!fiat_pasta_fp_equals_one(check)) {
```
Comment (mimoo) [inline pasta_fp.c:2157]: wouldn't zero also be ok?

**Review by mimoo — APPROVED** [review]
LGTM! Left mostly minor nits, besides me not understanding how you can `from_montgomery` without a `to_mongtomery` first. But if this works then it must be me.

**Pull request merged**